File: android/src/main/java/com/cloudcare/ft/mobile/sdk/tracker/reactnative/FTTraceModule.java
package com.cloudcare.ft.mobile.sdk.tracker.reactnative;

import com.cloudcare.ft.mobile.sdk.tracker.reactnative.utils.ReactNativeUtils;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableNativeMap;
import com.ft.sdk.FTSdk;
import com.ft.sdk.FTTraceConfig;
import com.ft.sdk.FTTraceManager;
import com.ft.sdk.TraceType;

import java.net.MalformedURLException;
import java.net.URISyntaxException;
import java.util.HashMap;
import java.util.Map;

public class FTTraceModule extends ReactContextBaseJavaModule {
  private static final String TAG = "FTTraceModule";

  public FTTraceModule(ReactApplicationContext reactContext) {
    super(reactContext);
  }

  @Override
  public String getName() {
    return "FTReactNativeTrace";
  }

  @ReactMethod
  public void setConfig(ReadableMap context, Promise promise) {
    Map<String, Object> map = context.toHashMap();
    Double sampleRate = (Double) map.get("sampleRate");
    Integer traceType = ReactNativeUtils.convertToNativeInt(map.get("traceType"));
    Boolean enableLinkRUMData = (Boolean) map.get("enableLinkRUMData");
    Boolean enableNativeAutoTrace = (Boolean) map.get("enableNativeAutoTrace");

    FTTraceConfig traceConfig = new FTTraceConfig();
    if (sampleRate != null) {
      traceConfig.setSamplingRate(sampleRate.floatValue());
    }

    if (traceType != null) {
      switch (traceType) {
        case 0:
          traceConfig.setTraceType(TraceType.DDTRACE);
          break;
        case 1:
          traceConfig.setTraceType(TraceType.ZIPKIN_MULTI_HEADER);
          break;
        case 2:
          traceConfig.setTraceType(TraceType.ZIPKIN_SINGLE_HEADER);
          break;
        case 3:
          traceConfig.setTraceType(TraceType.TRACEPARENT);
          break;
        case 4:
          traceConfig.setTraceType(TraceType.SKYWALKING);
          break;
        case 5:
          traceConfig.setTraceType(TraceType.JAEGER);
          break;
        default:
          traceConfig.setTraceType(TraceType.JAEGER);
          break;
      }
    }

    if (enableLinkRUMData != null) {
      traceConfig.setEnableLinkRUMData(enableLinkRUMData);
    }

    if (enableNativeAutoTrace != null) {
      traceConfig.setEnableAutoTrace(enableNativeAutoTrace);
    }

    FTSdk.initTraceWithConfig(traceConfig);
    //LogUtils.d("configCheck","traceConfig:"+new Gson().toJson(traceConfig));
    promise.resolve(null);
  }

  @ReactMethod
  public void getTraceHeaderFields(String url, String key, Promise promise) {
    HashMap<String, String> hashMap = null;
    try {
      if (key == null) {
        hashMap = FTTraceManager.get().getTraceHeader(url);
      } else {
        hashMap = FTTraceManager.get().getTraceHeader(key, url);
      }
    } catch (MalformedURLException e) {
      LogUtils.e(TAG, LogUtils.getStackTraceString(e));
    } catch (URISyntaxException e) {
      LogUtils.e(TAG, LogUtils.getStackTraceString(e));
    }

    WritableNativeMap map = new WritableNativeMap();
    if (hashMap != null) {
      for (Map.Entry<String, String> entry : hashMap.entrySet()) {
        map.putString(entry.getKey(), entry.getValue());
      }
    }
    promise.resolve(map);
  }
}

File: android/src/main/java/com/cloudcare/ft/mobile/sdk/tracker/reactnative/FTTraceModule.kt
package com.cloudcare.ft.mobile.sdk.tracker.reactnative

import com.cloudcare.ft.mobile.sdk.tracker.reactnative.utils.ReactNativeUtils
import com.facebook.react.bridge.*
import com.ft.sdk.FTSdk
import com.ft.sdk.FTTraceConfig
import com.ft.sdk.FTTraceManager
import com.ft.sdk.TraceType

class FTTraceModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {
  override fun getName(): String {
    return "FTReactNativeTrace"
  }

  @ReactMethod
  fun setConfig(context: ReadableMap, promise: Promise) {
    val map = context.toHashMap()
    val sampleRate = map["sampleRate"] as Float?
    val traceType = ReactNativeUtils.convertToNativeInt(map["traceType"])
    val enableLinkRUMData = map["enableLinkRUMData"] as Boolean?
    val enableNativeAutoTrace = map["enableNativeAutoTrace"] as Boolean?
//    val serviceName = map["serviceName"] as String?
    val traceConfig = FTTraceConfig()
    if (sampleRate != null) {
      traceConfig.samplingRate = sampleRate
    }

    if (traceType != null) {
      traceConfig.traceType = when (traceType) {
        0 -> TraceType.DDTRACE
        1 -> TraceType.ZIPKIN_MULTI_HEADER
        2 -> TraceType.ZIPKIN_SINGLE_HEADER
        3 -> TraceType.TRACEPARENT
        4 -> TraceType.SKYWALKING
        5 -> TraceType.JAEGER
        else -> TraceType.JAEGER
      }
    }

    if (enableLinkRUMData != null) {
      traceConfig.isEnableLinkRUMData = enableLinkRUMData
    }

    if (enableNativeAutoTrace != null) {
      traceConfig.isEnableAutoTrace = enableNativeAutoTrace
    }

//    if (serviceName != null) {
//      traceConfig.serviceName = serviceName
//    }

    FTSdk.initTraceWithConfig(traceConfig)

    promise.resolve(null)
  }

//  @ReactMethod
//  fun addTrace(
//    key: String,
//    context: ReadableMap, promise: Promise
//  ) {
//    val map = context.toHashMap()
//    val requestHeader = map["requestHeader"] as HashMap<String, String>
//    val httpMethod = map["httpMethod"] as String
//    val statusCode = ReactNativeUtils.convertToNativeInt(map["statusCode"])
//    val responseHeader = map["responseHeader"] as HashMap<String, String>?
//    val errorMsg = map["errorMsg"] as String?
//    FTTraceManager.get().addTrace(key, httpMethod, requestHeader, responseHeader,
//      statusCode ?: 0, errorMsg)
//    promise.resolve(null)
//  }

  @ReactMethod
  fun getTraceHeader(key: String, url: String, promise: Promise) {
    val hashMap = FTTraceManager.get().getTraceHeader(key, url)
    val map = WritableNativeMap()
    hashMap.forEach {
      map.putString(it.key, it.value)
    }
    promise.resolve(map)
  }
}
